### ISSUE TITLE
An error occurs when compiling to Android using cmake

### ISSUE THREAD

**Issue by leamus** [libsdl-org/SDL#8947]
Under windows, I compile the latest sdl3 with NDK23 and CMake3.28.1 with the following command: 

$cmake.. -G "Unix Makefiles" -DCMAKE_TOOLCHAIN_FILE="%NDK_HOME%/build/cmake/android.toolchain.cmake" -DANDROID_ABI="arm64-v8a" -DANDROID_PLATFORM="latest", 
$make

error message after completing immediately: 

[92%] Linking CXX shared library libSDL3.so
ld: error: undefined symbol: ACameraManager_openCamera
>>> referenced by SDL_android_video_capture.c:250 (D:/Documents/Desktop/Pleafles/_Yes/_Programming/_SDL/SDL/src/video/androidSDL_android_video_capture.c:250),
...

how can I solve it?

**Comment by madebr** [comment]
I think you're configuring with an invalid [`ANDROID_PLATFORM`](https://developer.android.com/ndk/guides/cmake#android_platform)

**Comment by leamus** [comment]
> I think you're configuring with an invalid [`ANDROID_PLATFORM`](https://developer.android.com/ndk/guides/cmake#android_platform)

No,setting it to latest is effective as it automatically selects android-31

**Comment by leamus** [comment]
不过我已经解决了，在CMakeLists.txt的 加一句：
  if(ANDROID)
    target_link_libraries(SDL3-shared PUBLIC camera2ndk mediandk) 即可

**Comment by leamus** [comment]
However, I have already resolved it by adding a sentence in CMakeLists.txt:

If (ANDROID)

Target_ Link_ Libraries (SDL3 shared PUBLIC camera 2ndk media) are sufficient

**Comment by leamus** [comment]
As a dynamic link library used in my project, there are still some issues (many Java native functions are not registered, I will solve them later)

**Comment by madebr** [comment]
> > I think you're configuring with an invalid [`ANDROID_PLATFORM`](https://developer.android.com/ndk/guides/cmake#android_platform)
> 
> No,setting it to latest is effective as it automatically selects android-31

You're right!

> However, I have already resolved it by adding a sentence in CMakeLists.txt:
> 
> If (ANDROID)
> 
> Target_ Link_ Libraries (SDL3 shared PUBLIC camera 2ndk media) are sufficient

Weird, I use this toolchain and can link `SDL3-shared` without issues:
```cmake
set(ANDROID_PLATFORM latest CACHE STRING "Android platform version" FORCE)
set(ANDROID_ABI arm64-v8a CACHE STRING "Android ABI" FORCE)
include(/path/to/Android/Sdk/ndk/23.1.7779620/build/cmake/android.toolchain.cmake)
```

But you're probably right, see these comments in the android video capture source:
https://github.com/libsdl-org/SDL/blob/f6f9468d8408604aaf6299668d1a456402cf240b/src/video/android/SDL_android_video_capture.c#L34-L38

Does this fix the link error? (I cannot reproduce)
```patch
--- a/Android.mk
+++ b/Android.mk
@@ -78,7 +78,7 @@
 
 LOCAL_CXXFLAGS += -std=gnu++11
 
-LOCAL_LDLIBS := -ldl -lGLESv1_CM -lGLESv2 -lOpenSLES -llog -landroid
+LOCAL_LDLIBS := -lcamera2ndk -ldl -lGLESv1_CM -lGLESv2 -lmediandk -lOpenSLES -llog -landroid
 
 LOCAL_LDFLAGS := -Wl,--no-undefined -Wl,--version-script=$(LOCAL_PATH)/src/dynapi/SDL_dynapi.sym
 
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1329,6 +1329,10 @@
     set(HAVE_CLOCK_GETTIME 1)
   endif()
 
+  if(SDL_VIDEO_CAPTURE)
+    sdl_link_dependency(videocapture LIBS camera2ndk mediandk)
+  endif()
+
   if(NOT SDL_DISABLE_ANDROID_JAR)
     find_package(Java)
     find_package(SdlAndroidPlatform MODULE)
```

**Comment by madebr** [comment]
> As a dynamic link library used in my project, there are still some issues (many Java native functions are not registered, I will solve them later)

`libSDL3.so` provides `JNI_OnLoad` which should bind all native Java methods. Please let us know what errors you see, and also how you eventually solved them.

Many thanks!

**Comment by leamus** [comment]
> --- a/Android.mk
+++ b/Android.mk
@@ -78,7 +78,7 @@
 
 LOCAL_CXXFLAGS += -std=gnu++11
 
-LOCAL_LDLIBS := -ldl -lGLESv1_CM -lGLESv2 -lOpenSLES -llog -landroid
+LOCAL_LDLIBS := -lcamera2ndk -ldl -lGLESv1_CM -lGLESv2 -lmediandk -lOpenSLES -llog -landroid
 
 LOCAL_LDFLAGS := -Wl,--no-undefined -Wl,--version-script=$(LOCAL_PATH)/src/dynapi/SDL_dynapi.sym
 
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -1329,6 +1329,10 @@
     set(HAVE_CLOCK_GETTIME 1)
   endif()
 
+  if(SDL_VIDEO_CAPTURE)
+    sdl_link_dependency(videocapture LIBS camera2ndk mediandk)
+  endif()
+
   if(NOT SDL_DISABLE_ANDROID_JAR)
     find_package(Java)
     find_package(SdlAndroidPlatform MODULE)

Successfully resolved the link issue，thanks

**Comment by leamus** [comment]
> > As a dynamic link library used in my project, there are still some issues (many Java native functions are not registered, I will solve them later)
> 
> `libSDL3.so`提供应绑定所有本机 Java 方法的方法。请让我们知道您看到了哪些错误，以及您最终是如何解决这些错误的。`JNI_OnLoad`
> 
> 非常感谢！

After using cmake, libSDL3. so, SDL3-3.0.0.jar, and SDL3-3.0.0-resources.jar will be generated. Last time, I only used libSDL3. so, which caused an issue. I estimate that I will need to use two more jar files to work properly. I will try it next

**Comment by madebr** [comment]
`SDL3-3.0.0.jar`contains the compiled java sources (class files), `SDL3-3.0.0-sources.jar` contains the sources such that the Android IDE does not have to decompile the class files.
These files are currently only useful for https://github.com/libsdl-org/SDL/issues/6315, or for building the Android tests with cmake, but without gradle.
Your old method of using the java sources should still work.

**Comment by leamus** [comment]
I'm very sorry for making a mistake. I forgot to delete the previous code when modifying the CMakeList file:
  if(ANDROID)
    target_link_libraries(SDL3-shared PUBLIC camera2ndk mediandk)
When I remove this code and add it according to your suggestion:
  if(SDL_VIDEO_CAPTURE)
    sdl_link_dependency(videocapture LIBS camera2ndk mediandk)
  endif()
It did not work and still reported an error during compilation:
[ 92%] Linking CXX shared library libSDL3.so
ld: error: unable to find library -lcamera2ndk
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
mingw32-make[2]: *** [CMakeFiles/SDL3-shared.dir/build.make:3712: libSDL3.so] Error 1
mingw32-make[1]: *** [CMakeFiles/Makefile2:92: CMakeFiles/SDL3-shared.dir/all] Error 2
make: *** [Makefile:156: all] Error 2

**Comment by leamus** [comment]
Yes, I copied “SDL3-3.0.0.jar” to the libs directory and all the issues have been resolved. Thank you very much

**Comment by madebr** [comment]
Nothing has changed in the SDL git repo. Did you apply my patch in your fork or did the problem disappear?

I don't think we can "just link" to camera2ndk and mediandk: the linker cannot find them with (some) ndk's.

**Comment by leamus** [comment]
I have updated the local CMakeList.txt file according to your method, and it is indeed linked properly:
`if(SDL_VIDEO_CAPTURE)
sdl_link_dependency(videocapture LIBS camera2ndk mediandk)
endif()`

I don't know the reason, because I mainly develop under Windows and Android is only used for testing

**Comment by icculus** [comment]
This is resolved now that #8633 is merged.
